packages/loopring_v3/contracts/core/impl/libtransactions/AccountUpdateTransaction.sol

```diff
@@ -50,12 +57,17 @@ library AccountUpdateTransaction
     {
         // Read the account update
         AccountUpdate memory accountUpdate = readAccountUpdate(data, offset);
+        AccountUpdateAuxiliaryData memory auxData = abi.decode(auxiliaryData, (AccountUpdateAuxiliaryData));
+
+        // Check validUntil
+        require(block.timestamp < auxData.validUntil, "WITHDRAWAL_EXPIRED");
+        accountUpdate.validUntil = auxData.validUntil;
```

Comment:
Why need to double check validUntil unchain? In AccountUpdateCircuit, we already has constraint to check:
 // Validate
        requireValidUntil(pb, state.timestamp, validUntil.packed, NUM_BITS_TIMESTAMP, FMT(prefix, ".requireValidUntil")),

This constraint will ensure validUntil < circuit's timestamp. And we already check the circuit's timestamp to consistent with block's timestamp when processing block.

        // Validate timestamp
        uint32 inputTimestamp = _block.data.toUint32(offset);
        offset += 4;
        require(
            inputTimestamp > block.timestamp - ExchangeData.TIMESTAMP_HALF_WINDOW_SIZE_IN_SECONDS() &&
            inputTimestamp < block.timestamp + ExchangeData.TIMESTAMP_HALF_WINDOW_SIZE_IN_SECONDS(),
            "INVALID_TIMESTAMP"
        );

Reply:
This worked when `validUntil` was part of the DA, which made sure `validUntil` passed through those checks in the circuit. But I removed it from the DA to save some bytes, and now put it into the auxiliary data instead. This data can be freely chosen the the operator and is not validated by the circuit in any way. The `validUntil` in the circuit and the `validUntil` onchain are not related and could have different values (doesn't make sense to do so, but the operator could do that).

Reply:
So should we simply remove the logic from the circuit? Or it is used by for example, trades.

Reply:
It's still used in the most common cases where an EdDSA signature is used. In that case, the `validUntil` used in the circuit is directly approved by the user so can be used directly. And if it's not a conditional transaction no `validUntil` is ever checked onchain.

Reply:
So if we decide to check validUntil onchain, we should remove the requireValidUntil from off chain circuit. Otherwise it's double checked.

Reply:
It's never double checked, we only check `validUntil` onchain when an onchain signature is used because we cannot depend on the circuit check to be valid. The operator can use any `validUntil` in the circuit unless it is used in combination with an EdDSA signature, because `validUntil` is part of the data signed by the user. If there is no EdDSA signature the operator can put in any `validUntil` into the circuit because it is never authorized anywhere on layer 2 and because we don't put it into the DA there is no link between layer 1 <-> layer 2.